Python 2/3 fixes

rahulporuri opened the pull request:
- encode the enaml source data in test files on python 2.
- encode the args passed to `gc.linear_context` on Python 2.
- encode the JSON data and explicitly write to file on Python 3.
- wrap calls to `dict.keys` where necessary
- test package on python 2 and 3 on travis CI

jwiggins commented on install-edm-osx.sh:
```diff
@@ -0,0 +1,17 @@
+        curl -o "$EDM_INSTALLER_PATH" -L "https://package-data.enthought.com/edm/osx_x86_64/${EDM_MAJOR_MINOR}/${EDM_PACKAGE}"
+    fi
+
+    sudo installer -pkg "$EDM_INSTALLER_PATH" -target /
```
Why not target the home directory?

jwiggins commented on .travis.yml:
```diff
@@ -9,19 +9,29 @@ cache:
-  - export PATH="${HOME}/edm/bin:${PATH}"
+  - if [[ ${TRAVIS_OS_NAME} == "linux" ]] ; then sh -e /etc/init.d/xvfb start; fi
+  - if [[ ${TRAVIS_OS_NAME} == 'linux' ]]; then ./install-edm-linux.sh; export PATH="${HOME}/edm/bin:${PATH}"; fi
+  - if [[ ${TRAVIS_OS_NAME} == 'osx' ]]; then ./install-edm-osx.sh; export PATH="${PATH}:/usr/local/bin"; fi
```
Is there are particular reason why we use a different install location for macOS vs Linux?

rahulporuri replied:
Not sure. Never bothered asking that question before. These changes are based on the travis config in traits.https://github.com/enthought/traits/blob/master/.travis.yml

jwiggins commented on ensemble/ctf/editor.py:
```diff
@@ -209,7 +210,11 @@ def _draw_colors(self, color_nodes, gc):
+            if six.PY2:
+                args = (b'pad', b'userSpaceOnUse')
+            else:
+                args = ('pad', 'userSpaceOnUse')
```
Since this is in the rendering code path _and_ it never changes, can you move it out to a module-level constant?

rahulporuri replied:
corrected.

jwiggins replied:
Where?

rahulporuri replied:
sorry. i corrected it locally. haven't pushed yet.

jwiggins commented on ensemble/ctf/tests/test_editor.py:
```diff
@@ -33,7 +35,8 @@ def setUp(self):
 """
-
+        if six.PY2:
+            enaml_source = enaml_source.encode('utf-8')
```
Does enaml's parser really expect input in the whatever `str` maps to? Or can you pass utf-8 bytes on both platforms?

(we could also think about fixing this in `traits_enaml` to make life easier)

rahulporuri replied:
I can't pass utf-8 bytes to enaml's parser on Python 3. For more info, see https://github.com/enthought/ensemble/pull/66#issuecomment-493005807

jwiggins replied:
That's unfortunate... OK.

jwiggins commented on ensemble/ctf/utils.py:
```diff
@@ -35,7 +38,11 @@ def save_ctf(transfer_func, filename):
+        if six.PY2:
+            fp.write(data)
+        else:
+            fp.write(data.encode('utf-8'))
```
`json.dumps` [always produces `str` objects](https://docs.python.org/3.6/library/json.html#json.dump)([py2](https://docs.python.org/2.7/library/json.html#json.dump) can produce `str` or `unicode`)? _And they dropped the `encoding` kwarg in Python 3?_

Really? 🙄

jwiggins commented on ensemble/ctf/utils.py:
```diff
@@ -1,7 +1,10 @@
 from __future__ import unicode_literals
+
+from io import open
```
Why not use the builtin `open`?

rahulporuri replied:
Ahh. When I was initially debugging the issue with saving JSON data, I wanted to open the file with `encoding` kwarg, which isn't an option with the builtin `open` on Python 2. We don't need this anymore. Reverting to using the builtin `open`.

jwiggins commented on ensemble/volren/tests/test_volume_renderer.py:
```diff
@@ -74,8 +77,8 @@ def _example_volume_mask(self, volume_array):
-        half = (depth/2, height/2, width/2)
-        quarter = (depth/4, height/4, width/4)
+        half = (depth // 2, height // 2, width // 2)
+        quarter = (depth // 4, height // 4, width // 4)
```
You should probably add `division` to the `__future__` imports.

jwiggins commented on ensemble/ctf/window_function_component.py:
```diff
@@ -98,7 +98,7 @@ def values(self):
         xs = np.linspace(center - radius, center + radius, num_samples)
         ys = self.window_func(num_samples) * self.opacity
-        return zip(xs, ys)
+        return list(zip(xs, ys))
```
Is it actually a problem if this returns an iterator?

rahulporuri replied:
If we return an iterator here, we will need to wrap the call elsewher

See https://github.com/enthought/ensemble/blob/66b23c72227d880c28cb17615d3044242ac33511/ensemble/ctf/editor.py#L176

jwiggins replied:
Not really though. The code you linked to is using the return value of `values` in a way which is compatible with an iterator. (unless I'm missing something)

rahulporuri replied:
Ohh. yes. sorry. `alpha_nodes` is iterated over so we don't have to do `list(zip(...))`.
I guess my main problem is with the difference in behavior of `values` between the various `*Node` objects.

Reverting this change for now.

rahulporuri commented:
Regarding the `enaml_source` type that is expected by `enaml` on Python 2 and 3, see below

Python 2 : 

<details>

```
In [1]: from enaml.core.enaml_compiler import EnamlCompiler

In [2]: from enaml.core.parser import parse

In [3]: enaml_source = u"""
   ...: from enaml.widgets.api import MainWindow
   ...: from traits_enaml.widgets.enable_canvas import EnableCanvas
   ...: 
   ...: enamldef MainView(MainWindow):
   ...:     attr editor
   ...: 
   ...:     initial_size = (400, 100)
   ...: """

In [4]: enaml_ast = parse(enaml_source)

In [5]: EnamlCompiler.compile(enaml_ast, '__temp__')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-1173a853ec18> in <module>()
----> 1 EnamlCompiler.compile(enaml_ast, '__temp__')

/home/rporuri/.edm/envs/oculith/lib/python2.7/site-packages/enaml/core/enaml_compiler.pyc in compile(cls, node, filename)
    182         # Create the compiler and generate the code.
    183         compiler = cls(filename=filename)
--> 184         return compiler.visit(node)
    185 
    186     def visit_Module(self, node):

/home/rporuri/.edm/envs/oculith/lib/python2.7/site-packages/enaml/core/enaml_ast.pyc in visit(self, node, *args, **kwargs)
    349         self._node_stack.append(node)
    350         try:
--> 351             result = visitor(node, *args, **kwargs)
    352         finally:
    353             self._node_stack.pop()

/home/rporuri/.edm/envs/oculith/lib/python2.7/site-packages/enaml/core/enaml_compiler.pyc in visit_Module(self, node)
    198         # Populate the body of the module.
    199         for item in node.body:
--> 200             self.visit(item)
    201 
    202         # Delete the template map.

/home/rporuri/.edm/envs/oculith/lib/python2.7/site-packages/enaml/core/enaml_ast.pyc in visit(self, node, *args, **kwargs)
    349         self._node_stack.append(node)
    350         try:
--> 351             result = visitor(node, *args, **kwargs)
    352         finally:
    353             self._node_stack.pop()

/home/rporuri/.edm/envs/oculith/lib/python2.7/site-packages/enaml/core/enaml_compiler.pyc in visit_PythonModule(self, node)
    216         cg = self.code_generator
    217         cg.set_lineno(node.lineno)
--> 218         cg.insert_python_block(node.ast)
    219 
    220     def visit_EnamlDef(self, node):

/home/rporuri/.edm/envs/oculith/lib/python2.7/site-packages/enaml/core/code_generator.pyc in insert_python_block(self, pydata, trim)
    400 
    401         """
--> 402         code = compile(pydata, self.filename, mode='exec')
    403         bp_code = bp.Code.from_code(code).code
    404         if trim:  # skip SetLineno and ReturnValue

TypeError: AST identifier must be of type str

In [6]: enaml_ast = parse(enaml_source.encode('utf-8'))

In [7]: EnamlCompiler.compile(enaml_ast, '__temp__')
Out[7]: <code object  at 0x7fa451fb4d30, file "__temp__", line 1>
```
</details>

Python 3:

<details>

```
In [1]: from enaml.core.parser import parse

In [2]: from enaml.core.enaml_compiler import EnamlCompiler

In [3]: enaml_source = u"""
   ...: from enaml.widgets.api import MainWindow
   ...: from traits_enaml.widgets.enable_canvas import EnableCanvas
   ...: 
   ...: enamldef MainView(MainWindow):
   ...:     attr editor
   ...: 
   ...:     initial_size = (400, 100)
   ...: """

In [4]: enaml_ast = parse(enaml_source)

In [5]: EnamlCompiler.compile(enaml_ast, '__temp__')
Out[5]: <code object  at 0x7f236a90ca50, file "__temp__", line 1>

In [6]: enaml_ast = parse(enaml_source.encode('utf-8'))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-c03e8042b261> in <module>()
----> 1 enaml_ast = parse(enaml_source.encode('utf-8'))

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/__init__.py in parse(enaml_source, filename)
     46     # of the control of Ply.
     47     try:
---> 48         return _parser.parse(enaml_source, filename)
     49     except ParsingError as parse_error:
     50         raise parse_error()

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/base_parser.py in parse(self, source, filename)
    243         try:
    244             lexer = self.lexer(filename)
--> 245             return self.parser.parse(source, debug=0, lexer=lexer)
    246         except ParsingError as parse_error:
    247             raise parse_error()

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/ply/yacc.py in parse(self, input, lexer, debug, tracking, tokenfunc)
    329             return self.parseopt(input, lexer, debug, tracking, tokenfunc)
    330         else:
--> 331             return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
    332 
    333 

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/ply/yacc.py in parseopt_notrack(self, input, lexer, debug, tracking, tokenfunc)
   1047                 if not lookahead:
   1048                     if not lookaheadstack:
-> 1049                         lookahead = get_token()     # Get the next token
   1050                     else:
   1051                         lookahead = lookaheadstack.pop()

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/base_lexer.py in token(self)
    541     def token(self):
    542         try:
--> 543             tok = next(self.token_stream)
    544             return tok
    545         except StopIteration:

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/lexer3.py in analyse_async(self, token_stream)
     94         async_depth = 0
     95 
---> 96         for tok in token_stream:
     97 
     98             if seen_async_def and tok.type == 'INDENT':

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/base_lexer.py in add_endmarker(self, token_stream)
    780 
    781     def add_endmarker(self, token_stream):
--> 782         for tok in token_stream:
    783             yield tok
    784 

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/base_lexer.py in synthesize_indentation_tokens(self, token_stream)
    709         token = None
    710 
--> 711         for token in token_stream:
    712             # WS only occurs at the start of the line
    713             # There may be WS followed by NEWLINE so

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/base_lexer.py in annotate_indentation_state(self, token_stream)
    662         indent = NO_INDENT
    663 
--> 664         for token in token_stream:
    665             token.at_line_start = at_line_start
    666 

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/base_lexer.py in create_strings(self, token_stream)
    595 
    596     def create_strings(self, token_stream):
--> 597         for tok in token_stream:
    598             if not tok.type.startswith("STRING_START_"):
    599                 yield tok

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/enaml/core/parser/base_lexer.py in ensure_token_lexer(self, token_stream)
    589         # a lexer, since that is required by the error handlers.
    590         lexer = self.lexer
--> 591         for tok in token_stream:
    592             if getattr(tok, 'lexer', None) is None:
    593                 tok.lexer = lexer

/home/rporuri/.edm/envs/oculith3/lib/python3.6/site-packages/ply/lex.py in token(self)
    312         while lexpos < lexlen:
    313             # This code provides some short-circuit code for whitespace, tabs, and other ignored characters
--> 314             if lexdata[lexpos] in lexignore:
    315                 lexpos += 1
    316                 continue

TypeError: 'in <string>' requires string as left operand, not int
```

</details>

rahulporuri commented on ensemble/ctf/editor.py:
```diff
@@ -220,14 +227,14 @@ def _draw_histogram(self, gc):
         h_values = log_values * h
         bin_edges = bin_edges - bin_edges.min()
-        bin_edges *= w / bin_edges.max()
+        bin_edges *= w // bin_edges.max()
```
This is the only change i'm not sure about. I'm not sure if the `width` attr of an enable `Container` is always an int or if it can be a float.

rahulporuri replied:
I've reverted this change. The division isn't a floor division anymore.

rahulporuri commented on ensemble/ctf/utils.py:
```diff
@@ -1,15 +1,17 @@
     def func(pos):
         bounds = component.bounds
-        return tuple([clip_to_unit(z / size) for z, size in zip(pos, bounds)])
+        return tuple([clip_to_unit(z // size) for z, size in zip(pos, bounds)])
```
One more division that I am not sure of. @jwiggins do you know off the top of your head if `component.bounds` is always an int or if it is a float?

jwiggins replied:
Wrapping the answer in `clip_to_unit` seems to imply a value between 0 and 1.

rahulporuri replied:
Yup. I've reverted the floor division to normal division.

rahulporuri commented on ensemble/volren/volume_renderer.py:
```diff
@@ -156,7 +157,7 @@ def _set_volume_clip_planes(self):
             return
 
-        bounds = [b/CLIP_MAX for b in self.data.bounds]
+        bounds = [b//CLIP_MAX for b in self.data.bounds]
```
note : `CLIP_MAX` is an integer.

rahulporuri commented:
@jwiggins can you take another look at this PR? Except for the comment regd edm installation location, I've addressed all of your comments. If it's okay with you, I'd like to look at the edm install location issue separately and not address it in this PR.
